init

zurabreon opened the pull request:
Начало работы с проектом

zurabreon committed:
init

Начало работы с проектом

zurabreon committed:
wip

Реализовал начальные методы. Осталось доработать функционал

zurabreon committed:
feat

Реализовал получение токена и удаление токена с включением и отключением интеграции соответственно. Собственное api перенес в файл authSync

zurabreon committed:
docs

zurabreon commented:
Есть вопрос по заданию. При установки интеграции в теле запроса я получаю auth_code, который можно использовать 1 раз, чтобы получить новый токен соответственно, но мне еще нужно получить id. Получить его не проблема, но как тогда этот id получать, чтобы обратиться к самой api, если для запроса к amo я должен использовать либо токен, либо этот auth_code, но я его уже использовал для получения токена и он больше не действителен. Нужно ли тогда в этом задании делать проверку на post-запросы или достаточно того, что у меня токен сохраняется по id-аккаунта пользователя и удаляется, если включить/отключить интеграцию?

zurabreon committed:
refactor

Смог разобраться с тем как получать токен и после обрабатывать post-запросы

zurabreon commented:
Смог разобраться с проблемой

zurabreon committed:
refactor

Избавился от лишних файлов

aivanov-reon commented on config.ts:
```diff
@@ -0,0 +1,15 @@
+    CLIENT_ID: "e7b44b96-1d16-4437-b7ad-76dfa5080729",
+    CLIENT_SECRET: "mzRQyKozc4aeNfiNlzLYxiEgDnviSRpoTphYdW6VhuHhfR4LKyNv8c3NN5Jn5inf",
+    //AUTH_CODE живет 20 минут, при перезапуске скрипта нужно брать новый
+    AUTH_CODE: "def50200c92aa17441c078691f1474f3dbc7140ddc63e8c353a4ac01ea9d442c0f05f9be257894119e1556187c3ef75a530bd77e1da0e2160623b5fc6662dbe8994ec1a5a32bfed3b1ee97c7e160958ba4c359fe19d4c1eb60247091b69972132f34f0f69403d29582af73b9030c85322892b00387f1f4dbeef1be12d03fa306662e0fc521f1fca431e6aa5974cbe9f78373cbe79a60b2ab6fa2c4167441c6c6024c3759f11280b8ea7399eb7fac430a271c0900f7053bfe6f15b65f514efc7 ...
```
AUTH_CODE приходит при установке, по идее не должен использоваться

zurabreon replied:
Убрал

aivanov-reon commented on config.ts:
```diff
@@ -0,0 +1,15 @@
+    //AUTH_CODE живет 20 минут, при перезапуске скрипта нужно брать новый
+    AUTH_CODE: "def50200c92aa17441c078691f1474f3dbc7140ddc63e8c353a4ac01ea9d442c0f05f9be257894119e1556187c3ef75a530bd77e1da0e2160623b5fc6662dbe8994ec1a5a32bfed3b1ee97c7e160958ba4c359fe19d4c1eb60247091b69972132f34f0f69403d29582af73b9030c85322892b00387f1f4dbeef1be12d03fa306662e0fc521f1fca431e6aa5974cbe9f78373cbe79a60b2ab6fa2c4167441c6c6024c3759f11280b8ea7399eb7fac430a271c0900f7053bfe6f15b65f514efc7 ...
+    REDIRECT_URI: "https://b381-94-232-94-147.ngrok-free.app/login",
+    SUB_DOMAIN: "new1701950444",
```
SUB_DOMAIN приходит при установке, также не должен использоваться из констант

zurabreon replied:
Убрал

aivanov-reon commented on config.ts:
```diff
@@ -0,0 +1,15 @@
+    SUB_DOMAIN: "new1701950444",
+    // конфигурация сервера
+    PORT: '2000',
+    BOT_TOKEN: ""
```
BOT_TOKEN - тоже не нужно

zurabreon replied:
Убрал

aivanov-reon commented on index.ts:
```diff
@@ -0,0 +1,77 @@
+    res.status(200).send({message: "ok"});
+});
+
+app.post("/contact", async(req: Request<unknown, unknown, WebHook, RequestQuery>, res: Response) => {
+
+    const accountId: string = req.body.account.id;
+
+    const authCode = config.AUTH_CODE;
+    const subDomain = config.SUB_DOMAIN; 
+
+    const api = new AuthSync(subDomain, authCode, accountId);
+
+    console.log(await api.getContact(5239253));
+
+    res.status(200).send({message: "ok"});
+})
```
Вот это не понимаю зачем тебе нужно? 
У тебя нет задачи обрабатывать веб-хуки по контактам

zurabreon replied:
Здесь я хотел просто протестировать работоспособность приложения. Сможет ли корректно обработаться post-запрос используя токен, а не код авторизации.
Убрал.

aivanov-reon commented on index.ts:
```diff
@@ -0,0 +1,77 @@
+    
+    const accountId = req.query.account_id; 
+
+    fs.unlink(`./authclients/${accountId}_amo_token.json`, err => {
+        if(err) {
+            throw err;
+        }
+        mainLogger.debug('Token deleted successfully');
+    });
```
Все правильно, но эту функциональность можно вынести в твой класс authSync, чтобы он управлял сохранением и удалением токенов.
Сделай отдельный метод для этого.

zurabreon replied:
Вынес в класс authSync в метод deleteToken

aivanov-reon commented on authSync.ts:
```diff
@@ -0,0 +1,149 @@
+                const token = res.data;
+                this.ACCESS_TOKEN = token.access_token;
+                this.REFRESH_TOKEN = token.refresh_token;
+                this.AMO_TOKEN_PATH = `./authclients/${await this.getAccountId()}_amo_token.json`
```
Так тоже можно, но не идеальное решение так как нужен доп. запрос. rоторый может провалиться в процессе сохранения токенов.
Попробуй сделать дополнительный метод, который будет доставать данные из jwt токена. 
Для этого тебе нужно установить библиотеку jwt-decode ( https://www.npmjs.com/package/jwt-decode )
и распарсить access_token. В нем есть account_id. Таким образом у тебя будет синхронная работа.

zurabreon replied:
Убрал доп.запрос, чтобы получать account_id и воспользовался jwt-decode. Решил не выносить это в отдельный метод, а передавать этот параметр в this.ACCOUNT_ID

aivanov-reon commented on index.ts:
```diff
@@ -0,0 +1,77 @@
+    res.status(200).send({message: "ok"});
+})
+
+app.get("/logout", async (req: Request<unknown, unknown, unknown, RequestQuery>, _res: Response) => {
```
В эту ветку тоже нужно добавить ответ. Сейчас его тут нет.

zurabreon replied:
Добавил ответ

aivanov-reon reviewed (commented):
Комментарии

zurabreon committed:
refactor

Избавился от доп.запроса, воспользовавшись библиотекой jwt-decode. Внес корректировки

aivanov-reon commented on index.ts:
```diff
@@ -0,0 +1,45 @@
+    mainLogger.debug('LOGOUT');
+    
+    const accountId = req.query.account_id; 
+    const api = new AuthSync('', '', '');
+
+    api.deleteToken(accountId);
```
Можно в api передавать аккаунт id и использовать его в методе.
Нужно в одном месте где-то оставить account_id 
В AuthSync он есть, но ты не передаешь его, но в deleteToken передаешь, хотя deleteToken может взять его из свойств класса.

zurabreon replied:
Исправил

aivanov-reon commented on api/amo.ts:
```diff
@@ -0,0 +1,172 @@
+                client_id: config.CLIENT_ID,
+                client_secret: config.CLIENT_SECRET,
+                grant_type: "refresh_token",
+                REFRESH_TOKEN: this.REFRESH_TOKEN,
```
REFRESH_TOKEN тут CapsLock вызовет ошибку. Поле должно быть refresh_token - это из документации amocrm

https://www.amocrm.ru/developers/content/oauth/step-by-step#:~:text=%D0%B4%D0%B5%D0%B9%D1%81%D1%82%D0%B2%D0%B8%D0%B5%20%D1%82%D0%BE%D0%BA%D0%B5%D0%BD%D0%B0%20%D0%B8%D1%81%D1%82%D0%B5%D0%BA%D0%B0%D0%B5%D1%82-,%D0%9F%D0%BE%D0%BB%D1%83%D1%87%D0%B5%D0%BD%D0%B8%D0%B5%20%D0%BD%D0%BE%D0%B2%D0%BE%D0%B3%D0%BE%20access%20token%20%D0%BF%D0%BE%20%D0%B5%D0%B3%D0%BE%20%D0%B8%D1%81%D1%82%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B8,-%D0%98%D0%B7%20%D0%BF%D1%80%D0%B5%D0%B4%D1%8B%D0%B4%D1%83%D1%89%D0%B5%D0%B3%D0%BE%20%D0%BF%D1%83%D0%BD%D0%BA%D1%82%D0%B0

zurabreon replied:
Не заметил. Исправил

aivanov-reon reviewed (commented):
Комменты

aivanov-reon commented:
Если что-то нужно тебе для тестов, то оставляй комментарии что это нужно для тестирования и напиши что проверял.

zurabreon committed:
refactor

Небольшие правки
